pkg/agent/controller/egress/ip_scheduler.go

```diff
@@ -0,0 +1,286 @@
+	if !shouldScheduleEgress(oldEgress) && !shouldScheduleEgress(curEgress) {
+		return
+	}
+	if oldEgress.Spec.EgressIP == curEgress.Spec.EgressIP && oldEgress.Spec.ExternalIPPool == curEgress.Spec.ExternalIPPool {
+		return
+	}
+	c.queue.Add(curEgress.Name)
+	klog.V(2).InfoS("Egress UPDATE event triggered Egress IP scheduling", "egress", klog.KObj(curEgress))
+}
+
+// deleteEgress processes Egress DELETE events.
+func (c *egressIPScheduler) deleteEgress(obj interface{}) {
+	egress, ok := obj.(*crdv1a2.Egress)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Received unexpected object: %v", obj)
+			return
+		}
+		egress, ok = deletedState.Obj.(*crdv1a2.Egress)
+		if !ok {
+			klog.Errorf("DeletedFinalStateUnknown contains non-Egress object: %v", deletedState.Obj)
+			return
+		}
+	}
+	if !shouldScheduleEgress(egress) {
+		return
+	}
+	c.queue.Add(egress.Name)
+	klog.V(2).InfoS("Egress DELETE event triggered Egress IP scheduling", "egress", klog.KObj(egress))
+}
+
+func (s *egressIPScheduler) Run(stopCh <-chan struct{}) {
+	klog.InfoS("Starting Egress IP scheduler")
+	defer klog.InfoS("Shutting down Egress IP scheduler")
+	defer s.queue.ShutDown()
+
+	if !cache.WaitForCacheSync(stopCh, s.egressListerSynced) {
+		return
+	}
+
+	// Schedule at least once even if there is no Egress to unblock clients waiting for HasScheduled to return true.
+	s.queue.Add(workItem)
+
+	go func() {
+		for {
+			obj, quit := s.queue.Get()
+			if quit {
+				return
+			}
+			s.schedule()
+			s.queue.Done(obj)
+		}
+	}()
+
+	<-stopCh
+}
+
+func (s *egressIPScheduler) HasScheduled() bool {
+	return s.scheduledOnce.Load()
+}
+
+func (s *egressIPScheduler) AddEventHandler(handler scheduleEventHandler) {
+	s.eventHandlers = append(s.eventHandlers, handler)
+}
+
+func (s *egressIPScheduler) GetEgressIPAndNode(egress string) (string, string, bool) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	result, exists := s.scheduleResults[egress]
+	if !exists {
+		return "", "", false
+	}
+	return result.ip, result.node, true
+}
+
+// EgressesByCreationTimestamp sorts a list of Egresses by creation timestamp.
+type EgressesByCreationTimestamp []*crdv1a2.Egress
+
+func (o EgressesByCreationTimestamp) Len() int      { return len(o) }
+func (o EgressesByCreationTimestamp) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
+func (o EgressesByCreationTimestamp) Less(i, j int) bool {
+	if o[i].CreationTimestamp.Equal(&o[j].CreationTimestamp) {
+		return o[i].Name < o[j].Name
+	}
+	return o[i].CreationTimestamp.Before(&o[j].CreationTimestamp)
+}
+
+func (s *egressIPScheduler) schedule() {
+	var egressesToUpdate []string
+
+	func() {
+		s.mutex.Lock()
+		defer s.mutex.Unlock()
+
+		newResults := map[string]*scheduleResult{}
+		nodeToIPs := map[string]sets.String{}
+		egresses, _ := s.egressLister.List(labels.Everything())
+		// Sort Egresses by creation timestamp to make the result deterministic and prioritize objected created earlier
```

Comment:
Question - it is still possible different agents get different lists for a period right? In that case, could different agents decide different IP assignment? Will that be corrected when all agents converge? Does it mean we may re-assign IPs?

Reply:
Yes, different agents may get differenet lists at a moment. But the diff should be likely the Egresses created most recently because they have been sent to some agents but haven't been sent to others.
As we sort Egresses by creation timestamp, Egresses created earilier will be prioritized and assigned to Nodes, and the results won't be affected by the Egresses created most recently. For instance, agent a1 receives Egress {e1, e2, e3}, agent a2 receives Egress {e1, e2}
* their schedult decision about e1 and e2 will be the same; 
* if a1 schedules e3 to itself, it will configure it to its interface, otherwise does nothing;
* when a2 receives e3, it should get the same result as a1, and configure e3 to its own interface or do nothing.

During the process no IP is re-assigned.

There could be also other cases causing different agents get different lists, e.g. Egress delete/update events. However, the behavior that one Egress's assignment may affect others only happens when Node capacity is reached. If the capacity is enough, each Egress's schedule is individual, and the consistent hash should guarantee the Egresses are distributed evenly. So in most cases when the Egress's number is not greater than Nodes * maxEgressIPsPerNode, there should be no IP re-assignning.

In all cases, all agents can get the same schedule results and correct IP assignment when their cache converge.

Reply:
My read is it is still possible that two agents decide different IP - Node assignment, when they get different Egress list, at Egress update/delete events? E.g. one gets {e1, e3}, one gets {e1, e2, e3}.

Good to add comments to describe the scenarios.

Reply:
Added comments to this function, PTAL